src/SpikeCore/SpikeCore/Irc/IrcClientBase.cs

```diff
@@ -1,24 +1,55 @@
 ﻿using System;
 using System.Collections.Generic;
+using System.Threading;
 using SpikeCore.Domain;
 
 namespace SpikeCore.Irc
 {
     public abstract class IrcClientBase: IIrcClient
     {
         private IEnumerable<string> _channelsToJoin;
+        protected string _host;
+        protected int _port;
+        protected string _nickname;
         protected bool _authenticate;
         protected string _password;
+        protected bool _userInitiatedDisconnect;
+        
         public WebHostCancellationTokenHolder WebHostCancellationTokenHolder { protected get; set; }
-
         public virtual Action<string> MessageReceived { get; set; }
         public virtual Action<PrivMessage> PrivMessageReceived { get; set; }
+        public virtual bool IsConnected { get; }
 
         public virtual void Connect(string host, int port, string nickname, IEnumerable<string> channelsToJoin, bool authenticate, string password)
         {            
             _channelsToJoin = channelsToJoin;
             _authenticate = authenticate;
             _password = password;
+
+            _host = host;
+            _port = port;
+            _nickname = nickname;
+            
+            Connect();
+        }
+
+        protected void HandleDisconnect(object sender, EventArgs e)
+        {      
+            UnwireEvents();
+            
+            if (!_userInitiatedDisconnect)
```

Comment:
Should this be inside the while loop with a return/break? That way if _userInitiatedDisconnect is set after the loop is entered it'll still exit correctly.

```
while()
{
    if (_userInitiatedDisconnect)
        break;

    [...]
}
```

Reply:
I'll go ahead and fold that back into the where clause - no need to break here. The method then becomes:

```
        protected void HandleDisconnect(object sender, EventArgs e)
        {      
            UnwireEvents();
            
            while (!_userInitiatedDisconnect && !IsConnected)
            {
                Connect();

                if (!IsConnected)
                {
                    // TODO - [kog@epiphanic.org 01/17/2019]: replace with a max retries + exponential backoff
                    Thread.Sleep(30000);
                }
            }
        }
```